### ISSUE TITLE
Exposing and implementing more core data structures and classes

### ISSUE THREAD

**Issue by Xrayez** [godot-extended-libraries/godot-ideas#11]
Related proposal: godotengine/godot-proposals#279.

# Problem
GDScript can be quite inefficient when dealing with non-trivial datatypes and structures. It takes at least an object to define some which could take a lot of memory consumption/leak problems etc. Using an ordinary array or a dictionary is simply not enough!

# Suggestion

There are a number of core data structures and classes which I'd like to see in Godot being built-in, and some which find some existing/future use cases.

### List

**Update**: This is implemented in [Goost](https://github.com/goostengine/goost), see [LinkedList documentation](https://goost.readthedocs.io/en/gd3/classes/class_linkedlist.html).

The advantage of this is having ability to pass around elements and see previous/next elements easily, which is somewhat difficult to do with an ordinary array. An `Element` would have to be exposed for this to work.

```gdscript
var list = List()
var element = list.push_back(value)
# ...
element = element.next
element = element.prev
```

See also godotengine/godot#7194.

### Set

When you need to ensure you only insert unique elements. It can be simulated with an array but you have to repeat code a lot, which basically boils down to this:

```gdscript
# Using array
var array = []
if not array.has(element):
    array.append(element)
var set = array # native
```
```gdscript
# Using dictionary
var dict = {}
dict[value] = null # or `true`
var set = dict.keys() # allocating array
```

### Graph

Currently present as a `GraphEdit` + `GraphNode` UI classes. It would be nice to have this class decoupled. See related proposal at godotengine/godot#15647. In fact, a doubly linked list could be represented as a [`PathGraph`](https://en.wikipedia.org/wiki/Path_graph) (as I think about it, it's a simple tree, which can be constructed with nodes, but still the scene tree is too general purpose for that).

### StringBuiler

I think this would be useful for plugins which need to write formatted output more easily, generate code etc (the way it's done in `VisualShader` for instance).

### [RingBuffer](https://en.wikipedia.org/wiki/Circular_buffer)

Stumbled upon this on accident, and I think it would be useful for input stuff, see godotengine/godot-proposals#100.

---

There are other structure and classes present in core, but I haven't found those to be performance critical, like `Pair`.

If you find some more use cases for existing core types which are not currently exposed to scripting, please describe it and I'll add it to the list here.

**Comment by willnationsdev** [comment]
Just as an FYI...

you can use a Dictionary<Variant, null> as a Set. It uses more memory than sit needs, but it otherwise accomplishes the same usability/speed task that a Set does.

I have a WIP List implementation in my fork of godot-next's `list` branch.

Creating a generic implementation of a graph is actually quite hard since their internal design is tied so deeply to the use case they target. You would need to define and expose a large variety of graph classes for them to be truly useful. I've already been here/done that with a `graph` branch in my Godot fork.

**Comment by Xrayez** [comment]
Yeah forgot to mention the usage of dictionary indeed. The performance drop could stem from allocating an array by having to use `dict.keys()` method down the road though, not sure.

**Comment by WilliamTambellini** [comment]
The core ring buffer header:
https://github.com/godotengine/godot/blob/master/core/ring_buffer.h
Is that idea/proposal been formally accepted by the godot owners : if anyone prepare a PullRequest in the master branch exposing ringbuffer, will it be reviewed/accepted/merged or is there any reason not to expose ringbuffer (for instance) ?
kind

**Comment by Xrayez** [comment]
@WilliamTambellini I don't think Godot core devs are even aware of this. One has to create a proposal at [GIP](https://github.com/godotengine/godot-proposals) first, but I think that the feature is a bit specific to be exposed to scripting (the existence of this idea here is to bring up possible use cases to justify making a formal proposal).

Talking about use cases, what do you need a ring buffer for personally?

**Comment by Xrayez** [comment]
So that you know, I've been working on a project which aims to complement Godot on the core level (and more than that), the project is [Goost](https://github.com/goostengine/goost) which is implemented as a C++ module, exposing core functionality to scripting as listed here could be likely worth it as almost every part of the engine is accessible through modules.

There was some ideas/discussion that `Graph` and related classes could be implemented in a general-purpose way (more or less), but I guess needs some design work up-front.

Of course we're basically limited to what the public API provides... So all those classes would have to be implemented inheriting `Object` for them to be exposed to scripting, I don't see other way.

**Comment by RaTcHeT302** [comment]
i wish we could just get rid of the gd script madness and turn c++ into a "just in time" compiled language instead so that we could move on with our lives and focus on making games instead of wasting our lives away with two terrible languages

at least if c++ was REALLY FAST when compiling, we would only have to deal with ONE terrible language xD

i really hate this mindset of, "it's too much hard work" or "we can't just throw it all away cuz X", but gd script is bad, c++ is bad, and i wish gd script was gone so that we could be focused on the c++ side of things, and then we could add support to just hot reload c++, and we would enjoy the benefits of a high performance language, with tons of hardware support, or we could just keep on torturing ourselves with... just another lame ass interpreted language system

idk i feel like putting more trust into actual computer scientists who put thousands of hours of effort into a language is more worthwhile than, dudes on the internet who made yet another pretend scripting language, i just want to cut the middle man and put some effort where it's worth and actually do something cool

i really want the engine to just get to a point where it's actually done, where things stop constantly breakng, but man i look at the billions of issues on the github tracker and there are like CONSTANTLY all kinds of random bugs involving gd script and it's so frustrating, it doesn't put me in the mood to learn how to use it at all, knowing how unstable it is, and i hate how the engine is more focused on adding flashy graphical features instead of focusing on the fact that the engine has a trillion bugs

i wish godot could focus on actually getting things to work, i think modern day software development is quite honestly retarded, everything is left broken and nothing is ever fixed, i wish they could just sit down and focus and actually finish one feature for once, but even vulkan is just sorta half assed, nothing is ever fully implemented, it's all just VERY half assed, it's all in the EVENTUALLY stage and IT'LL GET FIXED LATER (spoiler: it never gets fixed)

unity, unreal, godot, all these engines quite honestly frustrate me to no end, i just want to make a game without having to worry about the user having to deal with problems due to my own incompetence, to me that's unnacceptable and i find the state of modern day engines to be disgraceful, and i'm likely younger than everybody else here, and i just sorta feel like i'm insane for being the only guy who points out, just how ridiculous the lack of focus is

i mean go on the unity or unreal bug tracker, and look at some of the oldest bugs, the developers always say, OR THEY ARE EASY BUGS, WE'LL FIX THEM LATER

and yet 10 years later, they are still, just sitting there

boy don't you love modern day software development

i want this engine to be good, I WANT IT TO BE THE BEST OF THE BEST, but i don't feel like most people see that way, they just see a big meanie who uses BIG BAD WORDS, OH MY GOD HE SAID BAD THING, BAD PERSON, GO AWAY

it's depressing

**Comment by Xrayez** [comment]
@RaTcHeT302 I totally understand your frustration and even agree with general points you've made, but I guess that's the reality of software development...

You can read an interesting article about this in Leptonica (image processing library) documentation from a seasoned author which is pretty much blunt regarding this:

[What is “Well-Tested” C Code?](http://tpgit.github.io/UnOfficialLeptDocs/leptonica/testing-methods.html):

> I have seen estimates that commercial software has between 5 and 20 errors per 1000 lines of code. This may seem high, but many of these errors do not result in serious bugs. Taking the low end, one might guess that the code in the Leptonica library, which at present is about 100K lines, would have about 1000 errors. I hope it is much smaller, but obviously cannot prove it.

There are definitely areas which require well-tested code like medical software. [Games are soft real-time systems](https://stackoverflow.com/questions/17308956/differences-between-hard-real-time-soft-real-time-and-firm-real-time), so the amount of work is not justified to make it work reliably in 100% cases.

I just want to tell you that you're not crazy, but trying to change the world is also futile, the best thing you can do is to change ~~yourself~~ your project rather than the tools you use. 😛

> IT'LL GET FIXED LATER (spoiler: it never gets fixed)

Actually yes, but not by original authors, here's an example to keep this thread on topic: godotengine/godot#42013, I believe this was present way before Godot got open-sourced, even. 🙂

So, it's not a problem until it actually becomes a problem, and that's another issue with Godot development in my opinion (despite the regurgitated "Waiting for Godot" which could imply that the issues can be possibly foreseen before they appear). I'm not as experienced as other core developers in Godot, so perhaps I'm just venting in a similar way, but now I just tend to accept things as they are, else I'd just burn out.

**Comment by Xrayez** [comment]
Linked list implemented in the Goost module: goostengine/goost#12. Implementing a similar data structure in GDScript is possible but according to issues listed in godotengine/godot-proposals#1522, it may be difficult to do via GDScript, while also ensuring that the implementation is efficient enough, because in some cases using an `Array` would be faster (with some `resize` capacity hacks and moving indices around). 😉
